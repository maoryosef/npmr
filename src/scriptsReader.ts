import { join } from 'path';
import { cwd } from 'process';
import { readFileSync, existsSync } from 'fs';

interface PackageJson {
    scripts: Record<string, string>;
}

function readPackageJson(): PackageJson | null{
    const fileName = join(cwd(), 'package.json');
    
    if (existsSync(fileName)) {
        return JSON.parse(readFileSync(fileName, 'utf-8'));
    }
    
    return null;
}

export interface PackageJsonScript {
    name: string;
    cmd: string;
}

export const readScripts = (): PackageJsonScript[] =>
    Object.entries(readPackageJson()?.scripts || {})
        .reduce<PackageJsonScript[]>((acc, [name, cmd]) => [...acc, { name, cmd }], [])
