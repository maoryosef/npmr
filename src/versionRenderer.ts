import pkg from '../package.json'

export const render = (): void => {
    console.log(`
=============================================
    npmr version: ${pkg.version}             
=============================================
`)
}
