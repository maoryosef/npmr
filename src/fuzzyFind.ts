import fuzzysort from 'fuzzysort';
import { PackageJsonScript } from './scriptsReader';

const KEYS = ['cmd', 'name']
const HIGHLIGHT_SPLIT_CHAR = '\x00'
export interface Highlight {
  value: string;
  highlighted: boolean;
}
export interface SearchResult {
    value: PackageJsonScript;
    score: number;
    highlights: {
      cmd?: Highlight[];
      name?: Highlight[];
    }
}

const packageJsonScriptToSearchResult = (value: PackageJsonScript) : SearchResult =>
  ({value, score: 0, highlights: {}})

const toHighlights = (highlightedString: string | null) : Highlight[] | null => {
    if (!highlightedString) {
      return null;
    }
    
    const highlights: Highlight[] = []
    let highlighted = false;
    let value = ''    
    
    for (const c of highlightedString) {
      if (c === HIGHLIGHT_SPLIT_CHAR) {
        highlights.push({
          highlighted,
          value
        })
        
        value = ''
        highlighted = !highlighted
      } else {
        value += c
      }
    }
    
    if (value.length > 0) {
      highlights.push({
        highlighted: false,
        value
      })
    }
    
    return highlights
}

const getHighlights = (result: Fuzzysort.Result[], idx: number): Highlight[] | null =>
  toHighlights(fuzzysort.highlight(result[idx], HIGHLIGHT_SPLIT_CHAR, HIGHLIGHT_SPLIT_CHAR))


const fuzzySortResultToSearchResult = (result: Fuzzysort.KeyResult<PackageJsonScript>) : SearchResult =>
 ({
  value: result.obj,
  score: result.score,
  highlights: KEYS.reduce((obj, key, idx) =>
    ({ ...obj, [key]: getHighlights(result as unknown as Fuzzysort.Result[], idx) }), {}),
})

const doFuzzySort = (list: Array<PackageJsonScript>, query: string): SearchResult[] =>
  fuzzysort
    .go(query, list, {
      keys: KEYS,
      limit: 16, // Don't return more results than this (lower is faster)
      allowTypo: true, // (false is faster)
    })
    // @ts-ignore
    .sort((a, b) => a.score < b.score)
    .map(fuzzySortResultToSearchResult)

export default (list: Array<PackageJsonScript>, query: string = ''): SearchResult[] => 
    query.trim().length > 0 ? 
      doFuzzySort(list, query) :
      list.map(packageJsonScriptToSearchResult)