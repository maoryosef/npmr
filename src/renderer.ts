import chalk from 'chalk';
import figures from 'figures';

import { PackageJsonScript } from './scriptsReader';

export function renderValue(value: string, cursorPos: number): string {
    const prefix = chalk.cyan(`Search task${figures.pointer} `)
    
    if (value === '') {
        return prefix + value + chalk.dim.gray('Start typing to search...');
    }
    
    if (cursorPos === value.length) {
        return prefix + value + chalk.inverse(' ');
    }

    const firstChunk = value.slice(0, cursorPos);
    const corsurChar = value.slice(cursorPos, cursorPos + 1);
    const secondChunk = value.slice(cursorPos + 1);

    return prefix + firstChunk + chalk.inverse(corsurChar) + secondChunk;
}

export function renderScripts(scripts: PackageJsonScript[], selectedIdx: number): string {
    return scripts.map((script, idx) =>
        selectedIdx === idx ? 
            `${figures.play} ${chalk.bold(script.name)}\t ${chalk.white(script.cmd)}` :
            `  ${chalk.dim(script.name)}\t ${chalk.dim(script.cmd)}`
    ).join('\n');
}