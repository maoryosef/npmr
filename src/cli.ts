import logUpdate from 'log-update';
import InputHandler, { InputKey } from './input/inputHandler';
import * as terminal from './utils/terminal';
import {clamp} from './utils/math';
import * as renderer from './renderer';
import * as reader from './scriptsReader';
import { spawnSync } from 'child_process';
import fuzzyFind from './fuzzyFind';

const input = new InputHandler({ 
    stdin: process.stdin
})

terminal.clear();

const scripts = reader.readScripts();
let results = fuzzyFind(scripts)
let selectedIdx = 0;

function render(handler: InputHandler) {
    logUpdate([
        renderer.renderSearchString(handler.value, handler.cursorPos),
        renderer.renderScripts(results, selectedIdx),
    ].join('\n')
    )
}

render(input);

input.on('change', (handler, key) => {
    switch (key) {
        case InputKey.DOWN: selectedIdx++; break;
        case InputKey.UP: selectedIdx--; break;
    }
    
    results = fuzzyFind(scripts, handler.value);
    selectedIdx = clamp(selectedIdx, 0, results.length - 1);

    render(handler);
})

input.on('select', () => {
    terminal.clear();
    input.end();
    spawnSync('yarn', [scripts[selectedIdx].cmd], {stdio: 'inherit', shell: true});

    process.exit();
})

input.on('exit', () => {
    terminal.clear();
    input.end();
    process.exit();
})