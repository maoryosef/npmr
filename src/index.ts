import logUpdate from 'log-update';
import InputHandler, { InputKey } from './input/inputHandler';
import * as terminal from './utils/terminal';
import * as renderer from './renderer';
import * as reader from './scriptsReader';
import { spawnSync } from 'child_process';

const input = new InputHandler({ 
    stdin: process.stdin
})

terminal.clear();

const scripts = reader.readScripts();

let selectedIdx = 0;

function render(handler: InputHandler) {
    logUpdate([
        renderer.renderValue(handler.value, handler.cursorPos),
        renderer.renderScripts(scripts, selectedIdx),
    ].join('\n')
    )
}

render(input);

function clamp(value: number, min: number, max: number) {
    if (value < min) {
        return min;
    }
    
    if (value > max) {
        return max;
    }
    
    return value;
}

input.on('change', (handler, key) => {
    switch (key) {
        case InputKey.DOWN: selectedIdx++; break;
        case InputKey.UP: selectedIdx--; break;
    }
    
    selectedIdx = clamp(selectedIdx, 0, scripts.length - 1);

    render(handler);
})

input.on('select', () => {
    terminal.clear();
    input.end();
    spawnSync('yarn', [scripts[selectedIdx].cmd], {stdio: 'inherit', shell: true});

    process.exit();
})

input.on('exit', () => {
    terminal.clear();
    input.end();
    process.exit();
})