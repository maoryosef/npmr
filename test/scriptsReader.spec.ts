import {readScripts} from '../src/scriptsReader'
import {getFixtures, setFixture, beforeAndAfter} from './fixtures'

beforeAndAfter()

describe('scriptsReader', () => {
    test.each(getFixtures())('should read the scripts of "%s"', fixture => {

      setFixture(fixture)
      expect(readScripts()).toMatchSnapshot()
    });
})