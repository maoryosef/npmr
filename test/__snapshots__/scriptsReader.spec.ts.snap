// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scriptsReader should read the scripts of "emptyPackage" 1`] = `Array []`;

exports[`scriptsReader should read the scripts of "simplePackage" 1`] = `
Array [
  Object {
    "cmd": "tsc",
    "name": "build",
  },
  Object {
    "cmd": "webpack",
    "name": "build:webpack",
  },
  Object {
    "cmd": "tsc --watch",
    "name": "watch",
  },
]
`;
